### PR TITLE
LibWeb: Add input stepUp and stepDown functions

### PR DESCRIPTION
**Pull request by bplaat**
This adds the HTML input `stepUp()` and `stepDown()` functions, it was quite a lot to implement so I probably made some errors. Currently only really tested with `type="number"` because the datetime and range types are not really supported at the moment.

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 1]
```diff
@@ -1206,6 +1206,112 @@ String HTMLInputElement::covert_number_to_string(double input) const
...
+        return {};
+
+    // 2. Otherwise, if the attribute is absent, then the allowed value step is the default step multiplied by the step scale factor.
+    auto maybe_step_string = get_attribute(AttributeNames::step);
+    if (!maybe_step_string.has_value())
+        return default_step() * step_scale_factor();
+
+    // 3. Otherwise, if the attribute's value is an ASCII case-insensitive match for the string "any", then there is no allowed value step.
+    auto step_string = maybe_step_string.value();
+    if (step_string == "any")
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1281]: This doesn't look to be an ASCII case insensitive match

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 2]
```diff
@@ -1206,6 +1206,112 @@ String HTMLInputElement::covert_number_to_string(double input) const
...
+// https://html.spec.whatwg.org/multipage/input.html#concept-input-min-zero
+double HTMLInputElement::step_base() const
+{
+    // 1. If the element has a min content attribute, and the result of applying the algorithm to convert a string to a number to the value of the min content attribute is not an error, then return that result.
+    auto maybe_minimum = min();
+    if (maybe_minimum.has_value())
+        return maybe_minimum.value();
+
+    // 2. If the element has a value content attribute, and the result of applying the algorithm to convert a string to a number to the value of the value content attribute is not an error, then return that result.
+    auto maybe_value = convert_string_to_number(value().bytes_as_string_view());
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1302]: Do we need to explicitly call `bytes_as_string_view` here?

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 3]
```diff
@@ -1239,6 +1345,137 @@ WebIDL::ExceptionOr<void> HTMLInputElement::set_value_as_number(double value)
...
+    auto maybe_minimum = min();
+    auto maybe_maximum = max();
+    if (maybe_minimum.has_value() && maybe_maximum.has_value() && maybe_minimum.value() > maybe_maximum.value())
+        return {};
+
+    // 4. If the element has a minimum and a maximum and there is no value greater than or equal to the element's minimum andless than or equal to the element's maximum that, when subtracted from the step base, is an integral multiple of the allowed value step, then return.
+    // FIXME: Is this English?, implement this
+
+    // 5. If applying the algorithm to convert a string to a number to the string given by the element's value does not result in an error, then let value be the result of that algorithm. Otherwise, let value be zero.
+    double value = convert_string_to_number(this->value().bytes_as_string_view()).value_or(0);
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1371]: Similar here about explicit `bytes_as_string_view()`

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 4]
```diff
@@ -1239,6 +1345,137 @@ WebIDL::ExceptionOr<void> HTMLInputElement::set_value_as_number(double value)
...
```
Comment (shannonbooth) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1436]: Same here again

**Review by shannonbooth — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 5]
```diff
@@ -1199,17 +1331,148 @@ WebIDL::ExceptionOr<void> HTMLInputElement::set_value_as_number(double value)
...
-    m_value = MUST(String::number(value));
+    TRY(set_value(covert_number_to_string(value)));
+    return {};
+}
+
+// https://html.spec.whatwg.org/multipage/input.html#dom-input-stepup
+WebIDL::ExceptionOr<void> HTMLInputElement::step_up(long n)
+{
+    // 1. If the stepDown() and stepUp() methods do not apply, as defined for the input element's type attribute's current state, then throw an "InvalidStateError" DOMException.
+    if (type_state() != TypeAttributeState::Date && type_state() != TypeAttributeState::Month && type_state() != TypeAttributeState::Week && type_state() != TypeAttributeState::Time && type_state() != TypeAttributeState::LocalDateAndTime && type_state() != TypeAttributeState::Number && type_state() != TypeAttributeState::Range)
```
[the remark this answers is not in the document]
Comment (bplaat) [reply Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1352]: Hmm there a lot of different thing apply on input that we need to test https://html.spec.whatwg.org/multipage/input.html#concept-input-apply. I don't know how we can do that cleanly with a helper function so for now I will keep it the same.

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 6]
```diff
@@ -1180,15 +1180,41 @@ WebIDL::ExceptionOr<void> HTMLInputElement::set_size(unsigned value)
...
+    // https://html.spec.whatwg.org/multipage/input.html#number-state-(type=number):concept-input-value-string-number
+    if (type_state() == TypeAttributeState::Number) {
+        auto maybe_number = parse_floating_point_number(input);
+        if (maybe_number.has_value())
+            return maybe_number.value();
+        return {};
+    }
+
+    // FIXME: Implement more input types
+    VERIFY_NOT_REACHED();
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1195]: If we're going to crash, use `TODO` instead of `VERIFY_NOT_REACHED`. But it's generally less annoying to just add a `dbgln` with the type state to be implemented (and return an empty `Optional` in this case).

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 7]
```diff
@@ -1206,6 +1206,112 @@ String HTMLInputElement::covert_number_to_string(double input) const
     VERIFY_NOT_REACHED();
 }
 
+// https://html.spec.whatwg.org/multipage/input.html#attr-input-min
+Optional<double> HTMLInputElement::min() const
+{
+    // If the element has a min attribute, and the result of applying the algorithm to convert a string to a number to the value of the min attribute is a number, then that number is the element's minimum; otherwise, if the type attribute's current state defines a default minimum, then that is the minimum; otherwise, the element has no minimum.
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1212]: Please break very long lines into ~120 characters per line (here and elsewhere in this commit)

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 8]
```diff
@@ -1206,6 +1206,112 @@ String HTMLInputElement::covert_number_to_string(double input) const
...
+    return maybe_step.value() * step_scale_factor();
+}
+
+// https://html.spec.whatwg.org/multipage/input.html#concept-input-min-zero
+double HTMLInputElement::step_base() const
+{
+    // 1. If the element has a min content attribute, and the result of applying the algorithm to convert a string to a number to the value of the min content attribute is not an error, then return that result.
+    auto maybe_minimum = min();
+    if (maybe_minimum.has_value())
+        return maybe_minimum.value();
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1299]:
```c++
    if (auto minimum = min(); minimum.has_value())
        return *minimum;
```

and so on...

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 9]
```diff
@@ -1206,6 +1206,112 @@ String HTMLInputElement::covert_number_to_string(double input) const
...
+
+    // 2. If the element has a value content attribute, and the result of applying the algorithm to convert a string to a number to the value of the value content attribute is not an error, then return that result.
+    auto maybe_value = convert_string_to_number(value());
+    if (maybe_value.has_value())
+        return maybe_value.value();
+
+    // 3. If a default step base is defined for this element given its type attribute's state, then return it.
+    if (type_state() == TypeAttributeState::Week)
+        // The default step base is −259,200,000 (the start of week 1970-W01).
+        return -259200000;
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1309]: We consider this multi-line, thus should have `{ }` surrounding the body

Comment (trflynn89) [reply Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1309]: Also large numbers are easier to read with separators, at least for old dudes like me :)
```
-259'200'000
```

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 10]
```diff
@@ -1239,6 +1345,137 @@ WebIDL::ExceptionOr<void> HTMLInputElement::set_value_as_number(double value)
     return {};
 }
 
+// https://html.spec.whatwg.org/multipage/input.html#dom-input-stepup
+WebIDL::ExceptionOr<void> HTMLInputElement::step_up(long n)
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1349]: The spec states "The stepDown(n) and stepUp(n) methods, when invoked, must run the following algorithm:"

Let's not copy-paste the large function body for each of these, extract it to a common helper like `step_up_or_down` that `step_up` and `step_down` just invoke

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 11]
```diff
@@ -1239,6 +1345,137 @@ WebIDL::ExceptionOr<void> HTMLInputElement::set_value_as_number(double value)
...
+    double allowed_value_step = maybe_allowed_value_step.value();
+
+    // 3. If the element has a minimum and a maximum and the minimum is greater than the maximum, then return.
+    auto maybe_minimum = min();
+    auto maybe_maximum = max();
+    if (maybe_minimum.has_value() && maybe_maximum.has_value() && maybe_minimum.value() > maybe_maximum.value())
+        return {};
+
+    // 4. If the element has a minimum and a maximum and there is no value greater than or equal to the element's minimum andless than or equal to the element's maximum that, when subtracted from the step base, is an integral multiple of the allowed value step, then return.
+    // FIXME: Is this English?, implement this
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1368]: Just put a FIXME in front of the step number instead of a comment like this.
```c++
// FIXME: 4. ....
```

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 12]
```diff
@@ -1206,6 +1206,112 @@ String HTMLInputElement::covert_number_to_string(double input) const
...
+// https://html.spec.whatwg.org/multipage/input.html#attr-input-min
+Optional<double> HTMLInputElement::min() const
+{
+    // If the element has a min attribute, and the result of applying the algorithm to convert a string to a number to the value of the min attribute is a number, then that number is the element's minimum; otherwise, if the type attribute's current state defines a default minimum, then that is the minimum; otherwise, the element has no minimum.
+    auto maybe_min_string = get_attribute(HTML::AttributeNames::min);
+    if (maybe_min_string.has_value()) {
+        auto maybe_min = convert_string_to_number(maybe_min_string.value());
+        if (maybe_min.has_value())
+            return maybe_min.value();
+    }
```
[the remark this answers is not in the document]
Comment (bplaat) [reply Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1218]: Yes that looks much nicer

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp (hunk 10 above)
Comment (bplaat) [reply Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1349]: Good point, I need to stop with copying large parts of code and think a little bit more out of the box 😉

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 13]
```diff
@@ -1180,15 +1180,41 @@ WebIDL::ExceptionOr<void> HTMLInputElement::set_size(unsigned value)
...
 }
 
+// https://html.spec.whatwg.org/multipage/input.html#concept-input-value-string-number
+Optional<double> HTMLInputElement::convert_string_to_number(StringView input) const
+{
+    // https://html.spec.whatwg.org/multipage/input.html#number-state-(type=number):concept-input-value-string-number
+    if (type_state() == TypeAttributeState::Number) {
+        if (auto number = parse_floating_point_number(input); number.has_value())
+            return *number;
+        return {};
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1190]: Ah, I missed this before, but no need to even check the optional here:

```c++
if (type_state() == TypeAttributeState::Number)
    return parse_floating_point_number(input);
```

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 14]
```diff
@@ -1209,7 +1209,7 @@ String HTMLInputElement::covert_number_to_string(double input) const
 WebIDL::ExceptionOr<double> HTMLInputElement::value_as_number() const
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1316]: Re: the commit message, I think you mean "split off" (not of)? Let's reword it to be more clear, something like:
LibWeb: Move logic to check if valueAsNumber applies to its own function

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 15]
```diff
@@ -1205,6 +1205,113 @@ String HTMLInputElement::covert_number_to_string(double input) const
...
 
+// https://html.spec.whatwg.org/multipage/input.html#attr-input-min
+Optional<double> HTMLInputElement::min() const
+{
+    // If the element has a min attribute, and the result of applying the algorithm to convert a string to a number to
+    // the value of the min attribute is a number, then that number is the element's minimum; otherwise, if the type
+    // attribute's current state defines a default minimum, then that is the minimum; otherwise, the element has no minimum.
+    if (auto min_string = get_attribute(HTML::AttributeNames::min); min_string.has_value())
+        if (auto min = convert_string_to_number(*min_string); min.has_value())
+            return *min;
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1216]: Similar to your other PR, the outer `if` statement needs braces

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 16]
```diff
@@ -1205,6 +1205,113 @@ String HTMLInputElement::covert_number_to_string(double input) const
...
+
+// https://html.spec.whatwg.org/multipage/input.html#attr-input-max
+Optional<double> HTMLInputElement::max() const
+{
+    // If the element has a max attribute, and the result of applying the algorithm to convert a string to a number to the
+    // value of the max attribute is a number, then that number is the element's maximum; otherwise, if the type attribute's
+    // current state defines a default maximum, then that is the maximum; otherwise, the element has no maximum.
+    if (auto max_string = get_attribute(HTML::AttributeNames::max); max_string.has_value())
+        if (auto max = convert_string_to_number(*max_string); max.has_value())
+            return *max;
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1233]: Same (braces)

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 17]
```diff
@@ -1206,6 +1206,112 @@ String HTMLInputElement::covert_number_to_string(double input) const
...
+
+    // FIXME: Implement more input types
+    VERIFY_NOT_REACHED();
+}
+
+// https://html.spec.whatwg.org/multipage/input.html#concept-input-step
+Optional<double> HTMLInputElement::allowed_value_step() const
+{
+    // 1. If the attribute does not apply, then there is no allowed value step.
+    if (type_state() != TypeAttributeState::Date && type_state() != TypeAttributeState::Month && type_state() != TypeAttributeState::Week && type_state() != TypeAttributeState::Time && type_state() != TypeAttributeState::LocalDateAndTime && type_state() != TypeAttributeState::Number && type_state() != TypeAttributeState::Range)
```
[the remark this answers is not in the document]
Comment (trflynn89) [reply Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1268]: This comment is not resolved? Let's move this check to a helper function

**Review by trflynn89 — CHANGES_REQUESTED** [review]
This is looking much better! A few more comments.

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp [hunk 18]
```diff
@@ -1184,11 +1184,8 @@ WebIDL::ExceptionOr<void> HTMLInputElement::set_size(unsigned value)
 Optional<double> HTMLInputElement::convert_string_to_number(StringView input) const
 {
     // https://html.spec.whatwg.org/multipage/input.html#number-state-(type=number):concept-input-value-string-number
-    if (type_state() == TypeAttributeState::Number) {
-        if (auto number = parse_floating_point_number(input); number.has_value())
-            return *number;
-        return {};
-    }
+    if (type_state() == TypeAttributeState::Number)
+        return parse_floating_point_number(input);
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1188]: This fix was squashed into the wrong commit

**Review by trflynn89 — COMMENTED** [review]
You had one mis-squash, other than that LGTM!

File: Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp (hunk 18 above)
Comment (bplaat) [reply Userland/Libraries/LibWeb/HTML/HTMLInputElement.cpp:1188]: Oops

**Review by trflynn89 — APPROVED** [review]
